src/auth/src/credentials/idtoken.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::Result;
+use crate::token::Token;
+use std::future::Future;
+use std::sync::Arc;
+
+/// Represents an `IDTokenCredentials` used to obtain an ID token.
+///
+/// `IDTokenCredentials` provide a way to obtain OIDC ID tokens, which are
+/// commonly used for authenticating to services like Cloud Run or Identity-Aware
+/// Proxy (IAP). Unlike access tokens, ID tokens are not used to authorize access
+// to Google Cloud APIs but to verify the identity of a principal.
+///
+/// This struct serves as a wrapper around different credential types that can
+/// produce ID tokens, such as service accounts or metadata server credentials.
+#[derive(Clone, Debug)]
+pub struct IDTokenCredentials {
+    pub(crate) inner: Arc<dyn dynamic::IDTokenCredentialsProvider>,
+}
+
+impl<T> From<T> for IDTokenCredentials
+where
+    T: IDTokenCredentialsProvider + Send + Sync + 'static,
+{
+    fn from(value: T) -> Self {
+        Self {
+            inner: Arc::new(value),
+        }
+    }
+}
+
+impl IDTokenCredentials {
+    /// Asynchronously retrieves an ID token.
+    ///
+    /// This method will contact the underlying credential source to obtain a
+    /// fresh ID token if a valid one is not already cached.
+    pub async fn id_token(&self) -> Result<Token> {
+        self.inner.id_token().await
+    }
+}
+
+/// A trait for credential types that can provide OIDC ID tokens.
+///
+/// Implement this trait to create custom ID token providers. This is useful for
+/// testing or for integrating with authentication systems not natively
+/// supported by this library.
```

Comment:
I found the comment a bit awkward, I am not sure if this is better:

```suggestion
/// Implement this trait to create custom ID token providers.
/// For example, if you are working with an authentication system not
/// supported by this crate. Or if you are trying to write a test and need
/// to mock the existing `IDTokenCredentialsProvider` implementations.
```